dbms/src/Common/BitpackingPrimitives.h

```diff
@@ -0,0 +1,340 @@
+            if (values[i] > max_value)
+            {
+                max_value = values[i];
+            }
+
+            if constexpr (is_signed)
+            {
+                if (values[i] < min_value)
+                {
+                    min_value = values[i];
+                }
+            }
+        }
+        return findMinimumBitWidth<T, is_signed, round_to_next_byte>(min_value, max_value);
+    }
+
+    template <typename T, bool is_signed, bool round_to_next_byte = false>
+    constexpr static UInt8 findMinimumBitWidth(T min_value, T max_value)
+    {
+        UInt8 bitwidth;
+        T value;
+
+        if constexpr (is_signed)
+        {
+            if (min_value == std::numeric_limits<T>::min())
+            {
+                // handle special case of the minimal value, as it cannot be negated like all other values.
+                return sizeof(T) * 8;
+            }
+            else
+            {
+                value = std::max(static_cast<T>(-min_value), max_value);
+            }
+        }
+        else
+        {
+            value = max_value;
+        }
+
+        if (value == 0)
+        {
+            return 0;
+        }
+
+        if constexpr (is_signed)
+        {
+            bitwidth = 1;
+        }
+        else
+        {
+            bitwidth = 0;
+        }
+
+        while (value)
+        {
+            bitwidth++;
+            value >>= 1;
+        }
+
+        bitwidth = getEffectiveWidth<T>(bitwidth);
+
+        // Assert results are correct
+        if (bitwidth < sizeof(T) * 8 && bitwidth != 0)
+        {
+            if constexpr (is_signed)
+            {
+                RUNTIME_ASSERT(max_value <= (T(1) << (bitwidth - 1)) - 1);
+                RUNTIME_ASSERT(min_value >= (T(-1) * ((T(1) << (bitwidth - 1)) - 1) - 1));
+            }
+            else
+            {
+                RUNTIME_ASSERT(max_value <= (T(1) << (bitwidth)) - 1);
+            }
+        }
+        if constexpr (round_to_next_byte)
+        {
+            return (bitwidth / 8 + (bitwidth % 8 != 0)) * 8;
+        }
+        return bitwidth;
+    }
+
+    // Sign bit extension
+    template <typename T, typename TU = typename std::make_unsigned<T>::type>
+    static void signExtend(unsigned char * dst, UInt8 width)
+    {
+        T const mask = static_cast<T>(TU(1) << (width - 1));
+        for (size_t i = 0; i < BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE; ++i)
+        {
+            T value = unalignedLoad<T>(dst + i * sizeof(T));
+            value = static_cast<T>(value & ((TU(1) << width) - TU(1)));
+            T result = (value ^ mask) - mask;
+            unalignedStore<T>(dst + i * sizeof(T), result);
+        }
+    }
+
+    // Prevent compression at widths that are ineffective
+    template <typename T>
+    constexpr static UInt8 getEffectiveWidth(UInt8 width)
+    {
+        UInt8 bits_of_type = sizeof(T) * 8;
+        UInt8 type_size = sizeof(T);
+        if (width + type_size > bits_of_type)
```

Comment:
I don't really get the idea of this formula

Reply:
not a serious formula, mainly avoid some ineffective widths.

Reply:
If the compression ratio is less than 8/7, then it is ineffective, just copy is better.

Reply:
`width >= bits_of_type` is impossible. 

For example, `T` = UInt32, `width` = 31, uncompressed size = 1MB, then the compressed size is 0.97MB, the compression ration is quite low, there is no need to do bitpicking, just copy is much more effective.

Reply:
Any reasonable number is ok. `bits_of_type` is always equal to `type_size * 8`, here use `1/8` as threshold.

Reply:
Add some comments.